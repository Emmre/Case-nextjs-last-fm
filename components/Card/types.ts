export interface CardProps {
  data?: IAlbum | IArtist | ITrack
  isArtistList?: boolean
  isTopAlbum?: boolean
  isTopTrack?: boolean
}

interface IAlbum {
  name?: string
  artist?: string
  image?: IImage[]
  url?: string
  listeners?: string
  playcount?: string
}

interface IArtist {
  name?: string
  image?: IImage[]
  url?: string
  listeners?: string
  playcount?: string
}

interface ITrack {
  name?: string
  artist?: string
  image?: IImage[]
  url?: string
  listeners?: string
  playcount?: string
}

interface IImage {
  "#text": string
}
