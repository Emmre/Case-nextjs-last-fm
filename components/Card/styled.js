import styled from "styled-components"

export const CardStyled = styled.div`
  width: 100%;
  .card {
    gap: 16px;
    display: flex;
    background-color: #fff;
    border-radius: 4px;
    box-shadow: 0 0 10px rgba(0, 0, 0, 0.1);
    border: 1px solid #e6e6e6;
    padding: 20px;
    margin-bottom: 20px;
    margin-top: 20px;
    @media (max-width: 480px) {
      flex-direction: column;
    }

    &__image {
      object-fit: cover;
      margin-bottom: 10px;
    }

    &__content {
      font-size: 14px;
      line-height: 1.5;
      flex: 1;

      &__artist {
        font-weight: 600;
        border-bottom: 3px solid #000;
      }

      &__title {
        font-size: 18px;
        font-weight: 600;
        margin-bottom: 10px;
      }
    }

    &__stats {
      display: flex;
      flex-direction: column;
      justify-content: center;

      &__text {
        margin-bottom: 10px;

        &:last-child {
          margin-bottom: 0;

          a {
            color: #000;
            text-decoration: underline;

            &:hover {
              color: #000;
            }
          }
        }
      }
    }
  }
`

export const LinkStyled = styled.a``
