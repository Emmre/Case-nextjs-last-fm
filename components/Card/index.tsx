import { Fragment } from "react"
import { CardStyled, LinkStyled } from "./styled"
import { CardProps } from "./types"
import { useRouter } from "next/router"
import Image from "next/image"
const Card = ({
  data,
  isArtistList = false,
  isTopAlbum = false,
  isTopTrack = false,
  ...rest
}: CardProps) => {
  const { name = "", image = [], listeners = "", playcount = "" } = data || {}
  const router = useRouter()
  const { details } = router?.query || {}

  const urlToName = details?.toString().replaceAll("+", " ")
  const nameToUrl = name?.replaceAll(/ /g, "+")

  return (
    <CardStyled {...rest}>
      <LinkStyled
        href={`/artist/${nameToUrl}`}
        className={!isArtistList ? "disabled" : ""}
      >
        <div className='card'>
          <div className='card__image'>
            <Image
              src={
                image[1]?.["#text"] ||
                "https://lastfm.freetls.fastly.net/i/u/64s/2a96cbd8b46e442fc41c2b86b821562f.png"
              }
              alt={name}
              width={64}
              height={64}
            />
          </div>
          <div className='card__content'>
            {isArtistList && (
              <div className='card__content__artist'>
                <h3>Artist</h3>
              </div>
            )}
            {(isTopAlbum || isTopTrack) && (
              <div className='card__content__artist'>
                <h3>{name}</h3>
              </div>
            )}
            <h3 className='card__content__title'>{urlToName || name}</h3>
          </div>
          <div className='card__stats'>
            {isTopAlbum && <p>{playcount} play</p>}
            {isTopTrack && (
              <Fragment>
                <p className='card__text'>{listeners} listeners</p>
                <p className='card__text'>{playcount} play</p>
              </Fragment>
            )}
            {isArtistList && (
              <Fragment>
                <p className='card__text'>Listeners: {listeners}</p>
                <p className='card__text'>Playcount: {playcount}</p>
              </Fragment>
            )}
          </div>
        </div>
      </LinkStyled>
    </CardStyled>
  )
}

export default Card
