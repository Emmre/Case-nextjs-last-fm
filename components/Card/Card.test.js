import { render } from "@testing-library/react"
import "@testing-library/jest-dom"
import Card from "./index"

describe("CardComponent", () => {
  const props = {
    title: "Card Title",
    description: "Card Description",
    image: "https://via.placeholder.com/150",
  }

  it("Should render card component", () => {
    const { getByTestId } = render(<Card item={props} data-testid="card" />)
    expect(getByTestId("card")).toBeInTheDocument()
  })

  it("Should match snapshot", () => {
    const { container } = render(<Card item={props} data-testid="card" />)
    expect(container).toMatchSnapshot()
  })
})
