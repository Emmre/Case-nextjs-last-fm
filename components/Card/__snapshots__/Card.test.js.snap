// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardComponent Should match snapshot 1`] = `
<div>
  <div
    class="styled__CardStyled-sc-x26mvw-0 ilpskl"
    data-testid="card"
  >
    <a
      class="styled__LinkStyled-sc-x26mvw-1 disabled"
      href="/artist/"
    >
      <div
        class="card"
      >
        <div
          class="card__image"
        >
          <img
            alt=""
            src="https://lastfm.freetls.fastly.net/i/u/64s/2a96cbd8b46e442fc41c2b86b821562f.png"
          />
        </div>
        <div
          class="card__content"
        >
          <h3
            class="card__content__title"
          />
        </div>
        <div
          class="card__stats"
        />
      </div>
    </a>
  </div>
</div>
`;
